text/0000-grid-primitives.md

```diff
@@ -0,0 +1,679 @@
+    enum_value='white'
+)
+```
+
+Due to the use of protobuf, there is no valid default value, so not setting this
+value can either result in an error or can initialize the `enum_value` to one of
+the options in the property definition.  This detail is left to the
+smart-contract implementer.
+
+### Struct
+
+A struct value would be represented as follows:
+
+```
+PropertyDefinition(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_properties=[
+        PropertyDefinition(
+            name='speed',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+        PropertyDefinition(
+            name='duration',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+    ],
+    required=True
+)
+```
+
+An instance of the “shock" struct would be as follows:
+
+```
+PropertyValue(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_values=[
+        PropertyValue(
+            name='speed',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=500000),
+        PropertyValue(
+            name='duration',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=10000)
+        ])
+```
+
+The property value for a struct must contain all the struct values from the
+property definition, or it is invalid.  The defaults for the struct values
+themselves depend on their data types and/or the smart-contract implementer
+validation rules.
+
+### Latitude/Longitude
+
+A latitude/longitude (lat/long) value would be represented as follows:
+
+```
+PropertyDefintion(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    required=True
+)
+```
+
+A lat/long instance would be as follows:
+
+```
+PropertyValue(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    lat_long_value=LatLong(
+        latitude=44977753,
+        longitude=-93265015)
+)
+```
+
+Due to the use of protobuf, the default values for `LatLong` would be `(0, 0)`.
+While this is a valid lat/long, it could be used to indicate an error, depending
+on the choice of the smart-contract implementer.
+
+## Schema Example
+
+A complete object representation can be built from the property definition
+messages, and instances can be represented by constructing items with the
+property value messages.  
+
+Suppose there is a requirement to store different types of light bulbs as part
+of an application. A lightbulb may consist of the properties size, bulb type,
+energy rating, and color.
+
+We could define a Lightbulb schema as follows:
+
+```
+Schema(
+    name="Lightbulb",
+    description="Example Lightbulb schema",
+    owner = "philips001"
+    properties=[
+        PropertyDefinition(
+            name="size",
+            data_type=PropertyDefinition.DataType.NUMBER,
+            description="Lightbulb radius, in millimeters",
```

Comment:
Should properties `size` and `energy_rating` include the `number_exponent` field as well?

Reply:
Added.